### PR TITLE
Removes unused THCState code paths

### PR DESCRIPTION
**Pull request by mruberry**
To prepare THCState for refactoring into ATen, this PR removes unused THCState code paths. In particular, it:

- Removes the UVA Allocator
- Removes the THDefaultDeviceAllocator
- Respects the 1 BLAS and 1 sparse handle per device reality
- Removes kernel p2p access
- Removes setting p2p access
- Removes the GCHandler code path
- Removes many unused THCState_... functions
- Removes THCThreadLocal.h/.cpp

It does not change the preexisting external behavior of any remaining function.

### CODE REVIEW

File: aten/src/THC/THCAllocator.h [hunk 1]
```diff
@@ -4,7 +4,6 @@
 #include "THCGeneral.h"
 
 THC_API THAllocator* getTHCudaHostAllocator(void);
-THC_API THAllocator* getTHCUVAAllocator(void);
```
Comment (ezyang) [inline aten/src/THC/THCAllocator.h]: @nicolasvasilache Is tensor comprehensions using this still? I know you mentioned doing some experiments with this a while ago.

Comment (mruberry) [reply aten/src/THC/THCAllocator.h]: Maybe let's play it safe and keep the UVA Allocator for now?

**Comment by mruberry** [comment]
Just waiting for review.

File: aten/src/THC/THCAllocator.h (hunk 1 above)
Comment (ezyang) [reply aten/src/THC/THCAllocator.h]: No need, we can kill it.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Needs a rebase

File: aten/src/THC/THCAllocator.h (hunk 1 above)
Comment (mruberry) [reply aten/src/THC/THCAllocator.h]: Cool.

**Comment by mruberry** [comment]
Just a moment.

**Comment by mruberry** [comment]
Rebased. One Caffe2 test outstanding should have no issues. The merge with master was straightforward.

I think we're gtg.

**Pull request closed without merging**